hbase-common/src/main/java/org/apache/hadoop/hbase/nio/RefCnt.java

```diff
@@ -31,7 +34,10 @@
 @InterfaceAudience.Private
 public class RefCnt extends AbstractReferenceCounted {
 
-  private Recycler recycler = ByteBuffAllocator.NONE;
+  private static final ResourceLeakDetector<RefCnt> detector =
```

Comment:
I suppose it is not worth importing the ResourceLeakDetector code. How likely would netty change it? Pretty unlikely, I'd think. And we already use so many other netty types directly... Just thinking out loud.

Reply:
... and we get a single global control over this and netty itself via `-Dorg.apache.hbase.thirdparty.io.netty.leakDetection.level` . All good here.

Reply:
On the other hand, I am wondering if by having our private copy of `ResourceLeakDetector` was the only way by which @bbeaudreault was able to detect the leak while decompressing blocks.

Reply:
Looks like this is to be taken up by HBASE-27173 and not on this PR?

Reply:
I wrote up that issue before Andrew asked to add here. I'm happy to just close that issue if this lands here. Looks like this can land here easily, so will just close that one once this PR is merged.

@virajjasani I think Andrew's comment here was asking if we should copy the source code of ResourceLeakDetector into the hbase repo, to avoid having upstream changes break the leak detection for us. But I could be wrong.

Personally I don't think it's worth copying it in -- Looks like it's been around in netty for years and has been mostly stable, there was one minor improvement to it recently. I'd rather only copy it in if we have a reason to, like if we notice that they break it for us in later release. Maybe that's too reactionary though.

I'm happy to go with consensus here, but if we decide to copy it in I think we should maybe push this work to HBASE-27173 instead.

Reply:
> I think Andrew's comment here was asking if we should copy the source code of ResourceLeakDetector into the hbase repo, to avoid having upstream changes break the leak detection for us. But I could be wrong.

Ah, I see.

> Personally I don't think it's worth copying it in -- Looks like it's been around in netty for years and has been mostly stable, there was one minor improvement to it recently. I'd rather only copy it in if we have a reason to, like if we notice that they break it for us in later release. Maybe that's too reactionary though.

Fair enough.

> I wrote up that issue before Andrew asked to add here. I'm happy to just close that issue if this lands here. Looks like this can land here easily, so will just close that one once this PR is merged.

+1 for landing this change here.

> I could wrap it in a special `hbase.server.allocator.leakDetection.enabled` config

If this lands here, you would also like to consider this one right?

Reply:
Yea, though I wonder if it makes sense to leave as-is since it will be confusing to have 2 configs for the same feature. To Andrew's point above, this is already controlled via `-Dorg.apache.hbase.thirdparty.io.netty.leakDetection.level` (which also control's netty's own leak detection). 

Thoughts?

Reply:
Yeah makes sense, no need for separate config if same can be used for both netty and hbase together.
When I tried ADVANCED leak detection in HBASE-26708, I hardly got any useful hints, here is [the comment](https://issues.apache.org/jira/browse/HBASE-26708?focusedCommentId=17482669&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17482669).

One question: have you also increased `org.apache.hbase.thirdparty.io.netty.leakDetection.targetRecords` limit while debugging the whole issue?

Reply:
Yea, I upped it to like 6500. I also added a bunch of `touch("hint")` calls everywhere. It slowed down the region server a lot, but it was a test cluster so manageable. Not sure how much flexibility you have for something like that, maybe in the SASL layer.

Reply:
> Not sure how much flexibility you have for something like that, maybe in the SASL layer.

Yeah that's becoming more problematic only because of SASL layer. I upped target records to 100 and all the leak detected were only coming from Netty's EpollEventLoop. I even tried switching to `NioEventLoopGroup` but no luck, the leaks started coming from `NioEventLoop`.
Anyways, I am also going through Netty's code, not 100% confirmed but it seems these leak messages would appear all the time when `newDirectBuffer()` is called so it's pretty weird and I am not sure whether this leak is even leak or just a message that would appear regardless of the leak.
It's like this:
```
    @Override
    protected ByteBuf newDirectBuffer(int initialCapacity, int maxCapacity) {
        PoolThreadCache cache = threadCache.get();
        PoolArena<ByteBuffer> directArena = cache.directArena;

        final ByteBuf buf;
        if (directArena != null) {
            buf = directArena.allocate(cache, initialCapacity, maxCapacity);
        } else {
            buf = PlatformDependent.hasUnsafe() ?
                    UnsafeByteBufUtil.newUnsafeDirectByteBuf(this, initialCapacity, maxCapacity) :
                    new UnpooledDirectByteBuf(this, initialCapacity, maxCapacity);
        }

        return toLeakAwareBuffer(buf);
    }
```
And `toLeakAwareBuffer()` would report the leak for all the calls made to create new direct buffer 🤷

Reply:
Anyways, I am +1 for this PR, great work indeed :)